test/extensions/clusters/aggregate/cluster_test.cc

```diff
@@ -0,0 +1,221 @@
+    Upstream::HostConstSharedPtr target = lb_->chooseHost(nullptr);
+    EXPECT_EQ(host.get(), target.get());
+  }
+
+  EXPECT_CALL(primary_, loadBalancer()).WillRepeatedly(ReturnRef(primary_load_balancer_));
+  EXPECT_CALL(primary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(nullptr));
+  EXPECT_CALL(secondary_, loadBalancer()).WillRepeatedly(ReturnRef(secondary_load_balancer_));
+  EXPECT_CALL(secondary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(host));
+  for (int i = 66; i < 100; ++i) {
+    EXPECT_CALL(random_, random()).WillOnce(Return(i));
+    Upstream::HostConstSharedPtr target = lb_->chooseHost(nullptr);
+    EXPECT_EQ(host.get(), target.get());
+  }
+
+  // Set up the HostSet with 1 healthy, 1 degraded and 1 unhealthy.
+  Upstream::HostSharedPtr host1 = Upstream::makeTestHost(info_, "tcp://127.0.1.1:80");
+  host1->healthFlagSet(Upstream::HostImpl::HealthFlag::DEGRADED_ACTIVE_HC);
+  Upstream::HostSharedPtr host2 = Upstream::makeTestHost(info_, "tcp://127.0.1.2:80");
+  Upstream::HostSharedPtr host3 = Upstream::makeTestHost(info_, "tcp://127.0.1.3:80");
+  host2->healthFlagSet(Upstream::HostImpl::HealthFlag::FAILED_ACTIVE_HC);
+  host3->healthFlagSet(Upstream::HostImpl::HealthFlag::FAILED_ACTIVE_HC);
+  Upstream::HostSharedPtr host4 = Upstream::makeTestHost(info_, "tcp://127.0.1.4:80");
+
+  Upstream::HostVector hosts{host1, host2, host3, host4};
+  auto hosts_ptr = std::make_shared<Upstream::HostVector>(hosts);
+
+  primary_ps_.updateHosts(
+      0, Upstream::HostSetImpl::partitionHosts(hosts_ptr, Upstream::HostsPerLocalityImpl::empty()),
+      nullptr, hosts, {}, 100);
+
+  refreshLb();
+  // Health value:
+  // Cluster 1:
+  //     Priority 0: 25%
+  //     Priority 1: 33.3%
+  // Cluster 2:
+  //     Priority 0: 33.3%
+  //     Priority 1: 33.3%
+  EXPECT_CALL(primary_, loadBalancer()).WillRepeatedly(ReturnRef(primary_load_balancer_));
+  EXPECT_CALL(primary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(host));
+  EXPECT_CALL(secondary_, loadBalancer()).WillRepeatedly(ReturnRef(secondary_load_balancer_));
+  EXPECT_CALL(secondary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(nullptr));
+
+  for (int i = 0; i <= 57; ++i) {
+    EXPECT_CALL(random_, random()).WillOnce(Return(i));
+    Upstream::HostConstSharedPtr target = lb_->chooseHost(nullptr);
+    EXPECT_EQ(host.get(), target.get());
+  }
+
+  EXPECT_CALL(primary_, loadBalancer()).WillRepeatedly(ReturnRef(primary_load_balancer_));
+  EXPECT_CALL(primary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(nullptr));
+  EXPECT_CALL(secondary_, loadBalancer()).WillRepeatedly(ReturnRef(secondary_load_balancer_));
+  EXPECT_CALL(secondary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(host));
+  for (int i = 58; i < 100; ++i) {
+    EXPECT_CALL(random_, random()).WillOnce(Return(i));
+    Upstream::HostConstSharedPtr target = lb_->chooseHost(nullptr);
+    EXPECT_EQ(host.get(), target.get());
+  }
+}
```

Comment:
can you add a test case for when all hosts are unhealthy? as well as one in which panic mode is triggered

Reply:
Test cases added.